cpp/src/arrow/filesystem/mockfs.cc

```diff
@@ -242,8 +242,32 @@ class MockFSInputStream : public io::BufferReader {
     return metadata_;
   }
 
+  Result<int64_t> DoReadAt(int64_t position, int64_t nbytes, void* out) override {
+    RecordReadIo(position, nbytes);
+    return io::BufferReader::DoReadAt(position, nbytes, out);
+  }
+
+  Result<std::shared_ptr<Buffer>> DoReadAt(int64_t position, int64_t nbytes) override {
+    RecordReadIo(position, nbytes);
+    return io::BufferReader::DoReadAt(position, nbytes);
+  }
```

Comment:
@westonpace I follow your instruction to use the `MockFileSystem` to add unit tests. To do this, I tried several ways.
1) initially, I tried to call `GetRecordedReads` of `MockFSInputStream` in unit test to get the recorded IO, but I found this class is not public in the `mockfs.h` and I cannot use this class directly. 
2) After that, I tried to override the `ReadAt` virtual method in `io::BuffferReader` but I found `io::BufferReader` inherits from `internal::RandomAccessFileConcurrencyWrapper`, which marks `ReadAt` as `final`, which prevents me from overriding it.
Currently, I have to change the `protected` methods `DoReadAt` in `io::BufferReader` to be `virtual` so that it can be overridden for this kind of unit testing. But I am not entirely sure if this is desirable, please advise. Thanks.

Reply:
> While it's not merged yet, perhaps an approach like the one here will be easier? I don't think we need a full mock FileSystem, either - especially since the filesystem module isn't necessarily enabled. https://github.com/apache/arrow/pull/11535/files#diff-900c46995b5706697d6e4b010f610f1a1cf27d4d865afe48de0a800830ac676bL1708

Got it. It seems the simplest way is to modify this `TrackedRandomAccessFile` to add not only the number of read IO times but also the offset/length, but this PR is not merged yet, do you think if I can copy the `TrackedRandomAccessFile` into my PR and add the necessary change, and we can do some merge later?

Reply:
Sounds good to me. The same people are on both PRs so we should be aware.